File: DemoApp/DemoAppJava/app/src/main/java/co/optable/demoappjava/ui/PrebidBannerFragment.java
package co.optable.demoappjava.ui;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import co.optable.android_sdk.OptableIdentifier;
import co.optable.android_sdk.OptableResult;
import co.optable.android_sdk.OptableSDK;
import co.optable.android_sdk.OptableTargeting;
import co.optable.demoappjava.R;
import co.optable.demoappjava.TheApplication;
import com.google.android.gms.ads.AdListener;
import com.google.android.gms.ads.LoadAdError;
import com.google.android.gms.ads.admanager.AdManagerAdRequest;
import com.google.android.gms.ads.admanager.AdManagerAdView;
import com.google.common.collect.Lists;
import kotlin.Unit;
import org.jetbrains.annotations.NotNull;
import org.prebid.mobile.BannerAdUnit;
import org.prebid.mobile.TargetingParams;

import java.util.*;

public class PrebidBannerFragment extends Fragment {

    private static final String GAM_AD_UNIT_ID = "/21808260008/prebid_demo_app_original_api_banner";
    private static final String PREBID_CONFIG_ID = "prebid-demo-banner-320-50";
    private static final int WIDTH = 320;
    private static final int HEIGHT = 50;

    private AdManagerAdView adView;
    private BannerAdUnit prebidAdUnit;

    private ViewGroup adContainer;
    private TextView statusTextView;

    private OptableSDK optable;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View root = inflater.inflate(R.layout.fragment_prebid, container, false);
        initUi(root);
        optable = TheApplication.optable;
        return root;
    }

    private void initUi(View root) {
        statusTextView = root.findViewById(R.id.statusTextView);
        adContainer = root.findViewById(R.id.adContainer);

        root.findViewById(R.id.btnLoadBanner).setOnClickListener(view -> onClickLoadAd());
        root.findViewById(R.id.btnCachedBanner).setOnClickListener(view -> onClickCachedBanner());
        root.findViewById(R.id.btnClearCache).setOnClickListener(view -> onClickClearCache());
    }

    /**
     * Loads targeting data and then the GAM banner
     */
    private void onClickLoadAd() {
        statusTextView.setText("");

        ArrayList<OptableIdentifier.Email> ids = Lists.newArrayList(
                new OptableIdentifier.Email("test@test.com")
        );
        optable.targeting(ids, optableResult -> {
            OptableTargeting optableTargeting = null;
            if (optableResult instanceof OptableResult.Success<OptableTargeting> success) {
                optableTargeting = success.getData();
                changeStatusText("Targeting success");
                appendStatusText("GAM targeting keywords: " + optableTargeting.getGamTargetingKeywords());
                appendStatusText("OpenRTB JSON: " + optableTargeting.getOpenRtbJson());
                appendStatusText("Targeting data: " + optableTargeting.getTargetingData());
            } else if (optableResult instanceof OptableResult.Error<OptableTargeting> error) {
                changeStatusText("Targeting error: " + error.getMessage());
            }

            loadPrebidAd(optableTargeting);
            profile();
            witness();
        });
    }

    private void loadPrebidAd(@Nullable OptableTargeting optableTargeting) {
        AdManagerAdRequest.Builder adRequestBuilder = new AdManagerAdRequest.Builder();

        prebidAdUnit = new BannerAdUnit(PREBID_CONFIG_ID, WIDTH, HEIGHT);
        applyOptableToPrebid(optableTargeting);
        prebidAdUnit.fetchDemand(adRequestBuilder, resultCode -> {
            appendStatusText("Prebid ads loading status: " + resultCode.toString());
            loadGamAd(adRequestBuilder, optableTargeting);
        });
    }

    private void loadGamAd(AdManagerAdRequest.Builder requestBuilder, @Nullable OptableTargeting optableTargeting) {
        applyOptableToGam(requestBuilder, optableTargeting);

        adContainer.removeAllViews();

        adView = new AdManagerAdView(requireContext());
        adView.setAdUnitId(GAM_AD_UNIT_ID);
        adView.setAdSizes(new com.google.android.gms.ads.AdSize(WIDTH, HEIGHT));
        adView.setAdListener(new AdListener() {
            @Override
            public void onAdLoaded() {
                super.onAdLoaded();
                appendStatusText("Google ad loaded");
            }

            @Override
            public void onAdFailedToLoad(@NonNull @NotNull LoadAdError loadAdError) {
                appendStatusText("Google ad failed to load: " + loadAdError.getMessage());
            }
        });
        adView.loadAd(requestBuilder.build());

        adContainer.addView(adView);
    }

    /**
     * Loads targeting data from cache and then the GAM banner
     */
    private void onClickCachedBanner() {
        OptableTargeting targeting = optable.targetingFromCache();
        if (targeting != null) {
            changeStatusText("Targeting success");
            appendStatusText("GAM targeting keywords: " + targeting.getGamTargetingKeywords());
            appendStatusText("OpenRTB JSON: " + targeting.getOpenRtbJson());
            appendStatusText("Targeting data: " + targeting.getTargetingData());
        } else {
            changeStatusText("Targeting data cache empty.");
        }

        loadPrebidAd(targeting);
        profile();
        witness();
    }

    /**
     * Clears targeting data cache.
     */
    private void onClickClearCache() {
        statusTextView.setText("Clearing targeting data cache.\n\n");
        optable.targetingClearCache();
    }

    private void applyOptableToPrebid(@Nullable OptableTargeting optableResult) {
        if (optableResult == null) {
            TargetingParams.setGlobalOrtbConfig(null);
            return;
        }

        String openRtbJson = optableResult.getOpenRtbJson();
        if (openRtbJson != null) {
            TargetingParams.setGlobalOrtbConfig(openRtbJson);
        }
    }

    private void applyOptableToGam(AdManagerAdRequest.Builder builder, @Nullable OptableTargeting targeting) {
        if (targeting == null) return;

        Map<String, List<String>> audiences = targeting.getGamTargetingKeywords();
        if (audiences != null) {
            for (Map.Entry<String, List<String>> entry : audiences.entrySet()) {
                builder.addCustomTargeting(entry.getKey(), entry.getValue());
            }
        }
    }

    private void profile() {
        HashMap<String, Object> traits = new HashMap<>();
        traits.put("gender", "F");
        traits.put("age", 38);
        traits.put("hasAccount", true);
        traits.put("sampleFloat", 0.75);

        optable.profile(traits, "c:12", Set.of("c:id1", "c:id2"), result -> {
            if (result instanceof OptableResult.Success) {
                appendStatusText("Profile Success");
            } else if (result instanceof OptableResult.Error<OptableTargeting> error) {
                appendStatusText("Profile Error: " + error.getMessage());
            }
        });
    }

    private void witness() {
        HashMap<String, Object> eventProperties = new HashMap<>();
        eventProperties.put("exampleKey", "exampleValue");
        eventProperties.put("exampleKey2", 123);
        eventProperties.put("exampleKey3", false);

        optable.witness("GAMBannerFragment.loadAdButtonClicked", eventProperties, result -> {
            if (result instanceof OptableResult.Success) {
                appendStatusText("Witness Success");
            } else if (result instanceof OptableResult.Error<Unit> error) {
                appendStatusText("Witness Error: " + error.getMessage());
            }
        });
    }

    private void changeStatusText(@NonNull String message) {
        statusTextView.setText(message);
    }

    private void appendStatusText(@NonNull String message) {
        statusTextView.append("\n\n" + message);
    }

}

File: DemoApp/DemoAppKotlin/app/src/main/java/co/optable/androidsdkdemo/ui/PrebidBannerFragment.kt
/*
 * Copyright © 2020 Optable Technologies Inc. All rights reserved.
 * See LICENSE for details.
 */
package co.optable.androidsdkdemo.ui

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.TextView
import androidx.fragment.app.Fragment
import co.optable.android_sdk.OptableIdentifier
import co.optable.android_sdk.OptableResult
import co.optable.android_sdk.OptableSDK
import co.optable.android_sdk.OptableTargeting
import co.optable.androidsdkdemo.R
import co.optable.androidsdkdemo.TheApplication
import com.google.android.gms.ads.AdListener
import com.google.android.gms.ads.AdSize
import com.google.android.gms.ads.LoadAdError
import com.google.android.gms.ads.admanager.AdManagerAdRequest
import com.google.android.gms.ads.admanager.AdManagerAdView
import org.prebid.mobile.BannerAdUnit
import org.prebid.mobile.ResultCode
import org.prebid.mobile.TargetingParams

class PrebidBannerFragment : Fragment() {

    companion object {
        private const val GAM_AD_UNIT_ID = "/21808260008/prebid_demo_app_original_api_banner"
        private const val PREBID_CONFIG_ID = "prebid-demo-banner-320-50"
        private const val WIDTH = 320
        private const val HEIGHT = 50
    }

    private lateinit var adView: AdManagerAdView
    private lateinit var prebidAdUnit: BannerAdUnit

    private lateinit var optable: OptableSDK

    private lateinit var adContainer: ViewGroup
    private lateinit var statusTextView: TextView


    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View? {
        val root = inflater.inflate(R.layout.fragment_prebid, container, false)
        initUi(root)
        optable = TheApplication.optable
        return root
    }

    private fun initUi(root: View) {
        statusTextView = root.findViewById(R.id.statusTextView)
        adContainer = root.findViewById(R.id.adContainer)

        root.findViewById<Button>(R.id.btnLoadBanner).setOnClickListener { onClickLoadAd() }
        root.findViewById<Button>(R.id.btnCachedBanner).setOnClickListener { onClickCachedBanner() }
        root.findViewById<Button>(R.id.btnClearCache).setOnClickListener { onClickClearCache() }
    }

    /**
     * Loads targeting data and then the GAM banner.
     */
    private fun onClickLoadAd() {
        statusTextView.text = ""

        val ids = listOf(OptableIdentifier.Email("test@test.com"))
        optable.targeting(ids) { result ->
            val optableTargeting = when (result) {
                is OptableResult.Success<OptableTargeting> -> {
                    val targeting: OptableTargeting = result.data
                    changeStatusText("Targeting success")
                    appendStatusText("GAM targeting keywords: " + targeting.gamTargetingKeywords)
                    appendStatusText("OpenRTB JSON: " + targeting.openRtbJson)
                    appendStatusText("Targeting data: " + targeting.targetingData)
                    result.data
                }

                is OptableResult.Error -> {
                    changeStatusText("Targeting error: ${result.message}")
                    null
                }
            }

            loadPrebidAd(optableTargeting)

            profile()
            witness()
        }
    }

    private fun loadPrebidAd(optableTargeting: OptableTargeting?) {
        prebidAdUnit = BannerAdUnit(PREBID_CONFIG_ID, WIDTH, HEIGHT)
        applyOptableToPrebid(optableTargeting)

        val requestBuilder = AdManagerAdRequest.Builder()
        prebidAdUnit.fetchDemand(requestBuilder) { resultCode: ResultCode? ->
            appendStatusText("Prebid ads loading status: $resultCode")
            loadGamAd(requestBuilder, optableTargeting)
        }
    }

    private fun loadGamAd(requestBuilder: AdManagerAdRequest.Builder, optableTargeting: OptableTargeting?) {
        adContainer.removeAllViews()

        applyOptableToGam(requestBuilder, optableTargeting)

        adView = AdManagerAdView(requireContext())
        adView.setAdSizes(AdSize(WIDTH, HEIGHT))
        adView.adUnitId = GAM_AD_UNIT_ID
        adView.adListener = object : AdListener() {
            override fun onAdLoaded() {
                super.onAdLoaded()
                appendStatusText("Google ad loaded")
            }

            override fun onAdFailedToLoad(loadAdError: LoadAdError) {
                appendStatusText("Google ad failed to load: " + loadAdError.message)
            }
        }
        adView.loadAd(requestBuilder.build())

        adContainer.addView(adView)
    }


    /**
     * Loads cached targeting and then the GAM banner.
     */
    private fun onClickCachedBanner() {
        val targeting = optable.targetingFromCache()
        if (targeting != null) {
            changeStatusText("Targeting from cache success")
            appendStatusText("GAM targeting keywords: " + targeting.gamTargetingKeywords)
            appendStatusText("OpenRTB JSON: " + targeting.openRtbJson)
            appendStatusText("Targeting data: " + targeting.targetingData)
        } else {
            changeStatusText("Targeting data cache empty.")
        }

        loadPrebidAd(targeting)

        profile()
        witness()
    }

    /**
     * Clears the targeting data cache.
     */
    private fun onClickClearCache() {
        changeStatusText("Cleared targeting data cache.")
        optable.targetingClearCache()
    }


    private fun applyOptableToPrebid(targeting: OptableTargeting?) {
        if (targeting == null) {
            TargetingParams.setGlobalOrtbConfig(null)
            return
        }

        val openRtbJson = targeting.openRtbJson
        if (openRtbJson?.isNotBlank() == true) {
            TargetingParams.setGlobalOrtbConfig(openRtbJson)
        }
    }

    private fun applyOptableToGam(builder: AdManagerAdRequest.Builder, targeting: OptableTargeting?) {
        if (targeting == null) return

        val audiences = targeting.gamTargetingKeywords
        if (audiences != null) {
            for (entry in audiences.entries) {
                builder.addCustomTargeting(entry.key, entry.value)
            }
        }
    }

    private fun profile() {
        val traits = hashMapOf<String, Any>(
            "gender" to "F",
            "age" to 38,
            "hasAccount" to true,
            "sampleFloat" to 0.75,
        )

        optable.profile(traits, "c:12", setOf("c:id1", "c:id2")) { result ->
            when (result) {
                is OptableResult.Success -> {
                    appendStatusText("Profile success")
                }

                is OptableResult.Error -> {
                    appendStatusText("Profile error: ${result.message}")
                }
            }
        }
    }

    private fun witness() {
        val properties = hashMapOf<String, Any>("exampleKey" to "exampleValue", "anotherExample" to 123, "foo" to false)
        optable.witness("GAMBannerFragment.loadAdButtonClicked", properties) { result ->
            when (result) {
                is OptableResult.Success -> {
                    appendStatusText("Witness success")
                }

                is OptableResult.Error -> {
                    appendStatusText("Witness error: ${result.message}")
                }
            }
        }
    }


    private fun changeStatusText(message: String) {
        statusTextView.text = message
    }

    private fun appendStatusText(message: String) {
        statusTextView.append("\n\n$message")
    }

}
